hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestContainerReplicationUnderPlacementPolicy.java

```diff
@@ -0,0 +1,178 @@
+   * @param chosenNodes Provided node.
+   * @throws Exception
+   */
+  private void verifyReplicateCommandCount(int numProvidedNode,
+      int expectedCount, List<DatanodeDetails> chosenNodes) throws Exception {
+    List<DatanodeDetails> candidateNodes = new ArrayList<>();
+    if (chosenNodes != null) {
+      for (DatanodeDetails node : chosenNodes) {
+        candidateNodes.add(node);
+        numProvidedNode--;
+      }
+    }
+
+    for (int i = 0; i < numProvidedNode; i++) {
+      candidateNodes.add(randomDatanodeDetails());
+    }
+
+    conf = new OzoneConfiguration();
+    containerManager =
+        Mockito.mock(ContainerManager.class);
+    eventQueue = new EventQueue();
+    containerStateManager = new ContainerStateManager(conf);
+
+    Mockito.when(containerManager.getContainerIDs())
+        .thenAnswer(invocation -> containerStateManager.getAllContainerIDs());
+
+    Mockito.when(containerManager.getContainer(Mockito.any(ContainerID.class)))
+        .thenAnswer(invocation -> containerStateManager
+            .getContainer((ContainerID)invocation.getArguments()[0]));
+
+    Mockito.when(containerManager.getContainerReplicas(
+        Mockito.any(ContainerID.class)))
+        .thenAnswer(invocation -> containerStateManager
+            .getContainerReplicas((ContainerID)invocation.getArguments()[0]));
+
+    datanodeCommandHandler = new DatanodeCommandHandler();
+    eventQueue.addHandler(SCMEvents.DATANODE_COMMAND, datanodeCommandHandler);
+
+    SCMNodeMetric scmNodeMetric = Mockito.mock(SCMNodeMetric.class);
+    Mockito.when(scmNodeMetric.get()).thenReturn(new SCMNodeStat(1, 0, 1));
+
+    nodeManager =  Mockito.mock(NodeManager.class);
+    Mockito.when(nodeManager.getNodes(HddsProtos.NodeState.HEALTHY))
+        .thenReturn(candidateNodes);
+
+    Mockito.when(nodeManager.getNodeStat(Mockito.any(DatanodeDetails.class)))
+        .thenReturn(scmNodeMetric);
+
+    containerPlacementPolicy = new SCMContainerPlacementRandom(nodeManager,
+        conf, null, true, null);
+
+    replicationManager = new ReplicationManager(
+        new ReplicationManagerConfiguration(),
+        containerManager,
+        containerPlacementPolicy,
+        eventQueue,
+        new LockManager<>(conf));
+    replicationManager.start();
+    Thread.sleep(100L);
+
+    final ContainerInfo container = getContainer(LifeCycleState.CLOSED);
+    final ContainerID id = container.containerID();
+    containerStateManager.loadContainer(container);
+
+    // One replica in CLOSED state
+    DatanodeDetails node = (chosenNodes != null && !chosenNodes.isEmpty())
+        ? chosenNodes.get(0): randomDatanodeDetails();
+    final Set<ContainerReplica> replicas = getReplicas(id, State.CLOSED, node);
+
+    for (ContainerReplica replica : replicas) {
+      containerStateManager.updateContainerReplica(id, replica);
+    }
+
+    replicationManager.processContainersNow();
+    // Wait for EventQueue to call the event handler
+    Thread.sleep(100L);
```

Comment:
This as I understand is the part where the test replaces the replicas, and issues the command to the DataNodeCommandHandler, this can be also moved to a method called something like updateReplicas.